libs/common/.eslintrc.js

```diff
@@ -51,6 +51,7 @@ const baseRules = {
       consistent: true,
     },
   ],
+  'no-param-reassign': ['error', { props: true, ignorePropertyModificationsFor: ['draft', 'req', 'res'] }],
```

Comment:
just need approval from @abhi-markan on this

Reply:
Actually sorry @MarRobSoftwire - with fresh eyes - and forgetting previous comments, was this eslint change added for the `produce` function? If so, I think we can change the function and remove the need for this `no-param-reassign` rule entirely? E.g:

```ts
produce(existingDeal, (draft) => {
  const cloned = draft;  // <=====================

  propertiesToRemove.forEach((property) => {
    if (cloned[property]) {
      delete cloned[property];
    }
  });

  cloned._id = new ObjectId();
  cloned.version = getCurrentGefDealVersion();
  cloned.x = '...'

```

I don't think we should have `'draft'`  in the eslintrc, feels strange as it's very specific to cloning rather than a generic thing.

Reply:
This isnt specific to cloning but actually for the package `immer` - there's quite a few places that are using this package & disabling this rule. (see https://github.com/immerjs/immer/issues/189)

This rule is also violated whenever express middleware adds something to the `req` or `res` objects 
![image](https://github.com/user-attachments/assets/4c6d34f4-5e45-4325-97b5-688f75e11880)

Your example tony is still editing the parameters & feels like its hiding this. I quite like our linting to explictly allow in this case, rather than setting a precedent of using `// eslint-ignore` or reassigning the object.

Adding this rule in will allow us to remove lots of `// eslint-ignores` (https://ukef-dtfs.atlassian.net/browse/DTFS2-7603) which I'd love to get done

Reply:
My concern is, the rule is there for a reason - the example I provided is valid and is not editing the parameters; It creates a new instance.

Would be great to remove as many eslint ignore's as possible, but I don't think we should completely disable the rule.

I.e instead of `req.routePath` (not sure what this is used for), maybe using e.g `req.session` or `req.headers` would be better.

Reply:
Afaik, this rule is to stop mutability of objects. Immer uses a draft object that is mutated to create a resultant object that is immutable (I think it actually object.freezes). This rule can be relaxed for immer as a result.

Reply:
Yep, if we want to mutate anything on the `req` (e.g. session or headers) we need to disable the rule for these variables. The `res` object is needed for setting `res.locals` (e.g. gef-ui\server\middleware\csrf\index.js or gef-ui\server\middleware\sessionUserToLocals\index.js)

(I've just checked and we're disabling this rule anyway for `req` and `res` [by using Airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/best-practices.js#L237) )

Reply:
Ah nice -- So it's just for `draft` which is used in immer which is in discussion.

In which case is this controversial still as this is a key point to how [immer functions](https://immerjs.github.io/immer/) which is key for us `work with immutable state in a more convenient way`? 

This eslint rule is trying to achieve `immutable state` as well -- but in this case it's conflicting as immer requires it off to allow for an immutable state object to be created from another immutable state object through a convenient way. Having `draft` as an exception to the eslint rule shouldn't be an issue at all as it is only used by immer.

@ttbarnes any concerns?

Reply:
Sounds good to me.